designdocs/batched_messages.md

```diff
@@ -10,98 +10,198 @@
+consume a new message type.
+
+There are some upsides to deriving batching logically from existing structures.
+One is cost savings. Not having a batch structure means we don't need to worry
+about any on-chain computation associated with batch messages -- we simply look
+at the format and location on the parent schema and we can deduce whether the
+file is singular or plural.
 
 ### Alternatives and Rationale
-We discussed whether a batch message itself can be delegated, but this would have complicated things and we cannot come up with a use case for delegating batches. It also violates the idea of users delegating explicitly to every provider that performs a service for them, which is a fundamental value we want to apply to the network.
+We discussed whether a batch message itself can be delegated, but this would
+have complicated things and we cannot come up with a use case for delegating
+batches. It also violates the idea of users delegating explicitly to every
+provider that performs a service for them, which is a fundamental value we want
+to apply to the network.
 
 We discussed whether to allow URLs such as HTTP/HTTPS or other URLs and instead opted for content-addressable URIs (CIDv1) which can be resolved by some other service.  This allows us to put the file hash directly into a URI.  It reduces message storage because we don't have to include both a URL and a file hash. A file hash is necessary as a check against file tampering.
 
 We revisited the idea of whether it really is necessary to include a file size. We will be charging a premium for larger files, however, there will be per-byte discount for larger files in order to create an incentive for posting batches while reducing the incentive for announcers to allow spam. Although the processing and downloading time for enormous files also serves as a disincentive for spam, we feel it would not be sufficient.
 
-Despite the fact that announcers can lie abut the file size, the file_size parameter also serves as an on-chain declaration that not only allows consumers of batches to quickly discover if a batch announcer was honest, but the file requestor can know in advance when to stop requesting data.
+Despite the fact that announcers can lie abut the file size, the file_size
+parameter also serves as an on-chain declaration that not only allows consumers
+of batches to quickly discover if a batch announcer was honest, but the file
+requestor can know in advance when to stop requesting data.
+
+#### Message Payloads
+The document above describes payload location as an enumeration of on-chain and
+off-chain types. This allows us to encapsulate different information for both
+on-chain and IPFS payloads without overloading a single type with both sets of
+fields.
+
+However, if we discover that both on-chain and off-chain payloads are too
+complex to fit into a single enum, we could use traits to our advantage:
+
+```rust
+trait OnChainPayload {
```

Comment:
I don't know that we ever need to read this data on-chain, just write (or delete). The `get_messages_by_schema` rpc call does need to be able to build the correct response type however. Hmm... Mostly just thinking out loud